geoutils/geovector.py

```diff
@@ -320,3 +324,189 @@ def query(self: VectorType, expression: str, inplace: bool = False) -> VectorTyp
+        :param plot: Set to True to show intermediate plots, useful for understanding or debugging.
+
+        :returns: A Vector containing the buffered geometries.
+        """
+        # Dissolve all geometries into one
+        gdf = self.ds
+        merged = gdf.dissolve()
+
+        # Add buffer around geometries
+        merged_buffer = merged.buffer(buffer_size)
+
+        # Extract only the buffered area
+        buffer = merged_buffer.difference(merged)
+
+        # Crop Voronoi polygons to bound geometry and add missing polygons
+        bound_poly = gu.projtools.bounds2poly(gdf)
+        bound_poly = bound_poly.buffer(buffer_size)
+        voronoi_all = generate_voronoi_with_bounds(gdf, bound_poly)
+        if plot:
+            plt.figure(figsize=(16, 4))
+            ax1 = plt.subplot(141)
+            voronoi_all.plot(ax=ax1)
+            gdf.plot(fc='none', ec='k', ax=ax1)
+            ax1.set_title("Voronoi polygons, cropped")
+
+        # Extract Voronoi polygons only within the buffer area
+        voronoi_diff = voronoi_all.intersection(buffer.geometry[0])
+
+        # Join attributes of original geometries into the Voronoi polygons
+        voronoi_gdf = gpd.GeoDataFrame(geometry=voronoi_diff)
+        joined_voronoi = gpd.tools.sjoin(gdf, voronoi_gdf, how='right')
+
+        # Plot results -> some polygons are duplicated
+        if plot:
+            ax2 = plt.subplot(142, sharex=ax1, sharey=ax1)
+            joined_voronoi.plot(ax=ax2, column="index_left", alpha=0.5, ec='k')
+            gdf.plot(ax=ax2, column=gdf.index.values)
+            ax2.set_title("Buffer with duplicated polygons")
+
+        # Find non unique Voronoi polygons, and retain only first one
+        _, indexes = np.unique(joined_voronoi.index, return_index=True)
+        unique_voronoi = joined_voronoi.iloc[indexes]
+
+        # Plot results -> unique polygons only
+        if plot:
+            ax3 = plt.subplot(143, sharex=ax1, sharey=ax1)
+            unique_voronoi.plot(ax=ax3, column="index_left", alpha=0.5, ec='k')
+            gdf.plot(ax=ax3, column=gdf.index.values)
+            ax3.set_title("Buffer with unique polygons")
+
+        # Dissolve all polygons by original index
+        merged_voronoi = unique_voronoi.dissolve(by='index_left')
+
+        # Plot
+        if plot:
+            ax4 = plt.subplot(144, sharex=ax1, sharey=ax1)
+            gdf.plot(ax=ax4, column=gdf.index.values)
+            merged_voronoi.plot(column=merged_voronoi.index.values, ax=ax4, alpha=0.5)
+            ax4.set_title("Final buffer")
+            plt.show()
+
+        return gu.Vector(merged_voronoi)
+
+
+############################################
+# Additional stand-alone utility functions #
+############################################
+
+def extract_vertices(gdf: gpd.GeoDataFrame) -> list[list[tuple[float, float]]]:
+    """
+    Function to extract the exterior vertices of all shapes within a gpd.GeoDataFrame.
```

Comment:
Perhaps if the buffered polygon was just assigned all of the initial interior spaces, that could be a solution. But maybe the interior spaces should be buffered as well...

Reply:
Yes, that is certainly a good point too. As I said, that's what I wrote in a day of work, and I didn't want to get it lost, but since there are (probably) better implementations out there, I wonder if it's worth the effort...

Reply:
Maybe add it as an issue?